File: A1.hs
-- CSC254 Fall2019 HomeworkA1
-- Meiwen Zhou, Shreif Abdallah	
-- NetID: mzhou26, selsaid

-- [combine method]: helping function that functions as nested for loops 
-- (i,j) and returns a string list(result) of all the possible combinations
-- of 2 string lists (list1, list2). For 

combine:: Int -> Int -> [[Char]] -> [[Char]] -> [[Char]] -> [[Char]]
combine i j list1 list2 result 
    |i == length list1 = result 
    |j == length list2 = combine (i+1) 0 list1 list2 result
    |otherwise = combine i (j+1) list1 list2 (("("++(list1!!i)++"."++(list2!!j)++")"):result)

-- [tree method]: main function which takes one Int (n) and returns a 
-- list of possible n-node trees.
tree:: Int->[[Char]]
tree n 
    |n == 0 = [""]
    |n == 1 = ["(.)"]
    |otherwise =  (enumerator (n-1) 0 [])

-- [enumerator]: helping function that functions as a for loop which 
-- loops through all possible distributions of the left-child subtree 
-- tree(i) and right-child subtree tree(n-i-1), and returns a list of
-- strings of all possible binary trees that can be built from n nodes.
enumerator:: Int -> Int -> [[Char]] -> [[Char]]
enumerator i j result 
    |i < 0 = result 
    |otherwise = enumerator (i-1) (j+1) ((combine 0 0 (tree i) (tree j) [])++result)


-- [main]: PLEASE MODIFY NUMBER(N) IN PRINT((TREE N)) THEN RUN! :)
main = do
     print(tree 3)

File: A1.pl
# CSC254 Fall2019 HomeworkA1
# Meiwen Zhou, Shreif Abdallah 
# NetID: mzhou26, selsaid
# Thanks to Louis Jenkin's hint on Piazza 

# Base case tree(0) = ''
treeVal(0,'').

# Main function which reads number of nodes (N) and return a list of strings (T)
treeVal(N,T):-
    
    # Loop 1..N, functions as the outer loop that loops 
    between(1,N,X),

    # Recurssively call treeVal function which generates left subtree with I nodes
    I is N 
    treeVal(I,T1),

    # Recurssively call treeVal function which generates right subtree with J nodes 
    J is N-I-1,
    treeVal(J,T2),

    #combine left-child subtree with right-child subtree in this order: T = ( T1 . T2 )
    atom_concat('(',T1,A),
    atom_concat(A,'.',B),
    atom_concat(B,T2,C),
    atom_concat(D,')',T).